doc/tutorials/platformer/step_5.md

```diff
@@ -0,0 +1,329 @@
+first thing we need to do is make the game aware that collisions are going to occur using the
+`HasCollisionDetection` mixin.  Add that to `lib/ember_quest.dart` like:
+
+```dart
+class EmberQuestGame extends FlameGame
+    with HasCollisionDetection, HasKeyboardHandlerComponents {
+```
+
+Next, add the `CollisionCallbacks` mixin to `lib/actors/ember.dart` like:
+
+```dart
+class EmberPlayer extends SpriteAnimationComponent
+    with KeyboardHandler, CollisionCallbacks, HasGameRef<EmberQuestGame> {
+```
+
+If it did not auto-import, you will need the following:
+
+```dart
+import 'package:flame/collisions.dart';
+```
+
+Now add the following `onCollision` method:
+
+```dart
+@override
+void onCollision(Set<Vector2> intersectionPoints, PositionComponent other) {
+    if (other is GroundBlock || other is PlatformBlock) {
+        if (intersectionPoints.length == 2) {
+        // Calculate the collision normal and separation distance.
+        final mid = (intersectionPoints.elementAt(0) +
+                intersectionPoints.elementAt(1)) /
+            2;
+
+        final collisionNormal = absoluteCenter - mid;
+        final separationDistance = (size.x / 2) - collisionNormal.length;
+        collisionNormal.normalize();
+
+        // If collision normal is almost upwards,
+        // ember must be on ground.
+        if (fromAbove.dot(collisionNormal) > 0.9) {
+            isOnGround = true;
+        }
+
+        // Resolve collision by moving ember along
+        // collision normal by separation distance.
+        position += collisionNormal.scaled(separationDistance);
+        }
+    }
+
+    super.onCollision(intersectionPoints, other);
+}
+```
+
+You will need to import the following:
+
+```dart
+import '../objects/ground_block.dart';
+import '../objects/platform_block.dart';
+```
+
+As well as create these class variables:
+
+```dart
+  final Vector2 fromAbove = Vector2(0, -1);
+  bool isOnGround = false;
+```
+
+For the collisions to be activated for Ember, we need to add a `CircleHitbox`, so in the `onLoad`
+method, add the following:
+
+```dart
+add(
+    CircleHitbox(),
+);
+```
+
+Now that we have the basic collisions created, we can add gravity so Ember exists in a game world
+with very basic physics.  To do that, we need to create some more variables:
+
+```dart
+  final double gravity = 15;
+  final double jumpSpeed = 600;
+  final double terminalVelocity = 150;
+
+  bool hasJumped = false;
+```
+
+Now we can add Ember's ability to jump by adding the following to our `onKeyEvent` method:
+
+```dart
+hasJumped = keysPressed.contains(LogicalKeyboardKey.space);
+```
+
+Finally, in our `update` method we can tie this all together with:
+
+```dart
+// Apply basic gravity
+velocity.y += gravity;
+
+// Determine if ember has jumped
+if (hasJumped) {
+    if (isOnGround) {
+        velocity.y = -jumpSpeed;
+        isOnGround = false;
+    }
+    hasJumped = false;
+}
```

Comment:
Indentation is 2 spaces, not 4

Reply:
Same on all the other places too

Reply:
Oh, the old tab vs space debate...........at least my tab is set to 4 spaces.  Guess I will need to change that to 2, lol.

Reply:
Or run the code through `flutter format` 😉

Reply:
I have format on save set to true.  Maybe it's different or such. Will rerun it manually.

Reply:
It can't run in markdown files afaik

Reply:
Oh you know, that may be it, because you the settings are for dart not markdown and the formatter for markdown is different. Ok, will manually run and check it out.

Reply:
So just ran it, doesn't fix dart code in markdown and the formatters don't as well.  Has to be done manually.  I will adjust my tab to be 2 spaces.

Reply:
Fixed.